### ISSUE TITLE
IInvokedMethodListener and groupname

### ISSUE THREAD

**Issue by Jitu1888** [cbeust/testng#2692]
### TestNG Version
> 7.3.0
testNg.xml file I will add down (Runtime we are creating it)

### Expected behavior
it is traversing through all TestNG classes and finding the group and  it should not take the testNG class data members whose group name is not added.

### Actual behavior
it is traversing through all TestNG classes and finding the group but along with that, it is taking the testNG class data members whose group name is not added.

### Is the issue reproducible on a runner?
Its always

### Test case sample
> Wanted to know is it expected behavior or not

**Comment by Jitu1888** [comment]
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "https://testng.org/testng-1.0.dtd">
<suite thread-count="2" parallel="methods" data-provider-thread-count="2" name="My11Circle" verbose="2">
  <listeners>
    <listener class-name="com.appium.manager.AppiumParallelMethodTestListener"/>
    <listener class-name="com.appium.utils.RetryListener"/>
    <listener class-name="com.games24x7.e2e.core.listener.TestNGTestListener"/>
  </listeners>
  <test thread-count="2" parallel="methods" name="ATDTest" verbose="2">
    <parameter name="device" value=""/>
    <groups>
      <run>
        <include name="a-jitu"/>
      </run>
    </groups>
    <classes>
      <class name="com.games24x7.e2e.tests.functional.ft.FTSwitchTeamTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTMyTeamTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTDownloadTeamTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTChooseCaptainScreenRevampTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBundledContestTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTSocialLoginTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTTicketStoreTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBringAFriendTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTSportsTabLobbyTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTSubstitutionTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTDeleteMatchTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTLoyaltyPointsTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPrivateContestRevampTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBeatTheExpertCEFTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTAddCashLimitTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTAddCashTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBonusSummaryTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTDeepLinkTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTAddCashEnhancementsTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTMatchStartTimeTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPlayWithChampionsTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTTwoFactorAuthTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPracticeContestTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBonanzaLbTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPlaying11Test"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTCreateTeamFootBallTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBeatTheScoreTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTMarqueContestTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTLoginRegRevampTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTLoginTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBeatTheExpertTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTCreateTeamRevampTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPrivateContestTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPlaystoreRegLoginTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTNonIntrusiveTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPrivateContestWithdrawTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTCrossDeepLinkMECPrimaryTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTUpiIntentTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTContestTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTContestRevampTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTMeSectionTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBonusCashTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTMatchesTileTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTAddCashSPAChangesTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTTeamCreationTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTScoreBoardTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTDiscountedPromoCodeTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTContestPinningTest"/>
    </classes>
  </test> 
</suite>

**Comment by Jitu1888** [comment]
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "https://testng.org/testng-1.0.dtd">
<suite thread-count="2" parallel="methods" data-provider-thread-count="2" name="My11Circle" verbose="2">
  <listeners>
    <listener class-name="com.appium.manager.AppiumParallelMethodTestListener"/>
    <listener class-name="com.appium.utils.RetryListener"/>
    <listener class-name="com.games24x7.e2e.core.listener.TestNGTestListener"/>
  </listeners>
  <test thread-count="2" parallel="methods" name="ATDTest" verbose="2">
    <parameter name="device" value=""/>
    <groups>
      <run>
        <include name="a-jitu"/>
      </run>
    </groups>
    <classes>
      <class name="com.games24x7.e2e.tests.functional.ft.FTSwitchTeamTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTMyTeamTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTDownloadTeamTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTChooseCaptainScreenRevampTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBundledContestTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTSocialLoginTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTTicketStoreTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBringAFriendTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTSportsTabLobbyTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTSubstitutionTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTDeleteMatchTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTLoyaltyPointsTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPrivateContestRevampTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBeatTheExpertCEFTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTAddCashLimitTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTAddCashTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBonusSummaryTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTDeepLinkTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTAddCashEnhancementsTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTMatchStartTimeTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPlayWithChampionsTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTTwoFactorAuthTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPracticeContestTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBonanzaLbTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPlaying11Test"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTCreateTeamFootBallTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBeatTheScoreTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTMarqueContestTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTLoginRegRevampTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTLoginTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBeatTheExpertTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTCreateTeamRevampTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPrivateContestTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPlaystoreRegLoginTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTNonIntrusiveTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTPrivateContestWithdrawTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTCrossDeepLinkMECPrimaryTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTUpiIntentTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTContestTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTContestRevampTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTMeSectionTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTBonusCashTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTMatchesTileTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTAddCashSPAChangesTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTTeamCreationTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTScoreBoardTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTDiscountedPromoCodeTest"/>
      <class name="com.games24x7.e2e.tests.functional.ft.FTContestPinningTest"/>
    </classes>
  </test> 
</suite>

**Comment by Jitu1888** [comment]
[parallel.xml.zip](https://github.com/cbeust/testng/files/7725537/parallel.xml.zip)

**Comment by krmahadevan** [comment]
@Jitu1888 - Please edit the issue and add some context on what exactly is the problem. From what you have added, its not clear what is the issue you are facing.

The attached zip does not contain any information with respect to:

1. A simple standalone example that can be used to reproduce the problem.
2. No version information in terms of what TestNG version you are using. The latest released version of TestNG is `7.4.0`. Please ensure you use that and see if you are still able to reproduce the problem. Issues with previous versions cannot be fixed.

**Comment by Jitu1888** [comment]
1. I have attached the testNg.xml file.
2. There I have included a group name and added all the testNG classes.
3. Some testNG classes have some data members but the group name is not present in that TestNG class.
4. While executing the attached XML, it is taking the last TestNG class and instantiates the data members. For example
public class FTUpiIntentTest extends BaseTest {
    private int userBalance = 100;
    PlayerUtil playerUtil = new PlayerUtil(getTestExecutionContext());
    FTUpiIntentBL ftUpiIntentBL=new FTUpiIntentBL();
    private static String phonePe = "phone-pe.apk";
    private static String hdfcBank = "hdfc-mobilebanking.apk";
    private static String iciciBank = "icicibank-pockets.apk";
    private  String s3AutoBucketUrl = new PlayerUtil(getTestExecutionContext()).getPropValue(API.CONFIGS,"s3AutoBucketUrl");

    
   @Test(groups = { "a-upiIntent", "a-FtRegression", "i-upiIntent" })
    public void verifyListOfUPIAppInstalledTest(){

        HashMap<PlayerUtil.RegistrationDetails, String> userData = (HashMap<RegistrationDetails, String>) playerUtil.setupUserForUPITest();
        PlayerUtil.downloadApkFilesFromS3Bucket(phonePe, s3AutoBucketUrl + phonePe);
        ftUpiIntentBL.installPhonepeApp();
        new FTLoginBL().loginUsing(userData.get(PlayerUtil.RegistrationDetails.USERNAME),
                userData.get(PlayerUtil.RegistrationDetails.PASSWORD))
        .startFTDAddCash().verifyListOfUPIPaymentModes(userBalance);
        ftUpiIntentBL.removePhonepeApp();
    }

But in parallel.xml I have added  <include name="a-jitu"/>. So the above testNG class should be skipped and it should run the data members of that class right (PlayerUtil(getTestExecutionContext()).getPropValue(API.CONFIGS,"s3AutoBucketUrl");). Please correct me if I am wrong.

**Comment by Jitu1888** [comment]
@krmahadevan

**Comment by Jitu1888** [comment]
Please let me know if u need any more information @krmahadevan

**Comment by krmahadevan** [comment]
@Jitu1888 - Ok so if I understand your question correctly, you are basically expecting the following:

_I have "x" number of classes in my suite xml, and I have specified a particular group say "foo". You are expecting that TestNG should instantiate ONLY those classes out of the entire "x" classes, which satisfy the basic expectation of it having one or more methods that belong to the group "foo"_

Is this your expectation ? 

If my understanding of your expectation is correct, then AFAIK TestNG does not work that way. TestNG would instantiate all classes that you have specified but it would use the groups that you specified to filter out methods for execution.

So in a nutshell, groups filtering is meant to filter out what to execute and NOT to filter out what to instantiate and what to skip.

So all the classes are going to be instantiated all the time irrespective of whether it has any methods that belong or are absent from the selected group in question.

Let me know if that answers your question

**Comment by Jitu1888** [comment]
But with TestNG version 6.14.3: Behavior was different, there it did not instantiate the classes which were not specified with groups. Yesterday only I updated the testNG version then I faced some issues. @krmahadevan

**Comment by krmahadevan** [comment]
@Jitu1888 - In that case, please do the following:

Create a sample Maven/Gradle project with some code that can be used to reproduce the problem, along with the expected and actual outputs. 

Note: Please donot attach your full fledged tests that use selenium/appium/restassured etc., because we cannot run them. Also please ensure that the code in the sample project compiles.

**Comment by Jitu1888** [comment]
ok @krmahadevan give me some time, I will share a demo project with u. Thanks for you quick response

**Comment by krmahadevan** [comment]
@Jitu1888 - Sure thing no worries. Take your time. But please do help in sharing a proper reproducible sample so that its easy to dig into what broke where and see how we can fix it.

**Comment by krmahadevan** [comment]
No response from the user. Closing this issue. Please comment if the problem can be reproduced with a sample project using the latest released version (`7.6.1` as of today)
